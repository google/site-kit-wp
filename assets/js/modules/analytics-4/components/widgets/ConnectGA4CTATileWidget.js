/**
 * ConnectGA4CTATileWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * Internal dependencies
 */
import { useSelect } from 'googlesitekit-data';
import ConnectModuleCTATile from '../../../../components/KeyMetrics/ConnectModuleCTATile';
import useWidgetStateEffect from '../../../../googlesitekit/widgets/hooks/useWidgetStateEffect';
import {
	CORE_USER,
	keyMetricsGA4Widgets,
} from '../../../../googlesitekit/datastore/user/constants';
import Null from '../../../../components/Null';

const metadata = { moduleSlug: 'analytics-4' };

export default function ConnectGA4CTATileWidget( { Widget, widgetSlug } ) {
	const ga4DependantKeyMetrics = useSelect( ( select ) => {
		const keyMetrics = select( CORE_USER ).getKeyMetrics();

		if ( ! keyMetrics ) {
			return [];
		}

		return keyMetrics.filter( ( keyMetric ) =>
			keyMetricsGA4Widgets.includes( keyMetric )
		).length;
	} );

	// Only render the CTA if there are 3 or fewer GA4-dependent key metrics tiles.
	// Otherwise, render `Null` to hide the CTA.
	// If there are four GA4-dependent key metrics tiles, the `ConnectGA4CTAWidget` will be rendered instead.
	const ConnectModuleCTAComponent =
		ga4DependantKeyMetrics > 3 ? Null : ConnectModuleCTATile;

	useWidgetStateEffect( widgetSlug, ConnectModuleCTAComponent, metadata );

	// Note that we need to render the `Widget` component as a wrapper here so this component will display
	// correctly when used as a `FallbackComponent` for the `whenActive` HOC. Conversely, when `ConnectModuleCTATile`
	// is rendered as an `OverrideComponent` in `WidgetRenderer` (as a result of multiple adjacent widgets rendering
	// this component and thus sharing the same state), it is wrapped with a `Widget` component - the net
	// result being the ConnectModuleCTATile is correctly wrapped in a `Widget` in both cases.
	return (
		<Widget>
			<ConnectModuleCTATile { ...metadata } />
		</Widget>
	);
}

ConnectGA4CTATileWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
	widgetSlug: PropTypes.string.isRequired,
};
