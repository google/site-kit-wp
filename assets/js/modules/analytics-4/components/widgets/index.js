/**
 * GA4 Key Metrics widget components.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { default as ConnectGA4CTATileWidget } from './ConnectGA4CTATileWidget';
export { default as ConnectGA4CTAWidget } from './ConnectGA4CTAWidget';
export { default as EngagedTrafficSourceWidget } from './EngagedTrafficSourceWidget';
export { default as LeastEngagingPagesWidget } from './LeastEngagingPagesWidget';
export { default as MostEngagingPagesWidget } from './MostEngagingPagesWidget';
export { default as NewVisitorsWidget } from './NewVisitorsWidget';
export { default as PopularContentWidget } from './PopularContentWidget';
export { default as PopularProductsWidget } from './PopularProductsWidget';
export { default as ReturningVisitorsWidget } from './ReturningVisitorsWidget';
export { default as TopCitiesWidget } from './TopCitiesWidget';
export { default as TopCitiesDrivingLeadsWidget } from './TopCitiesDrivingLeadsWidget';
export { default as TopCitiesDrivingAddToCartWidget } from './TopCitiesDrivingAddToCartWidget';
export { default as TopCitiesDrivingPurchasesWidget } from './TopCitiesDrivingPurchasesWidget';
export { default as TopDeviceDrivingPurchasesWidget } from './TopDeviceDrivingPurchasesWidget';
export { default as TopCountriesWidget } from './TopCountriesWidget';
export { default as TopTrafficSourceWidget } from './TopTrafficSourceWidget';
export { default as TopTrafficSourceDrivingAddToCartWidget } from './TopTrafficSourceDrivingAddToCartWidget';
export { default as TopTrafficSourceDrivingLeadsWidget } from './TopTrafficSourceDrivingLeadsWidget';
export { default as TopConvertingTrafficSourceWidget } from './TopConvertingTrafficSourceWidget';
export { default as TopTrafficSourceDrivingPurchasesWidget } from './TopTrafficSourceDrivingPurchasesWidget';
export { default as PagesPerVisitWidget } from './PagesPerVisitWidget';
export { default as VisitLengthWidget } from './VisitLengthWidget';
export { default as TopReturningVisitorPages } from './TopReturningVisitorPages';
export { default as VisitsPerVisitorWidget } from './VisitsPerVisitorWidget';
export { default as TopRecentTrendingPagesWidget } from './TopRecentTrendingPagesWidget';
export { default as TopCategoriesWidget } from './TopCategoriesWidget';
export { default as TopPagesDrivingLeadsWidget } from './TopPagesDrivingLeadsWidget';
export { default as PopularAuthorsWidget } from './PopularAuthorsWidget';
