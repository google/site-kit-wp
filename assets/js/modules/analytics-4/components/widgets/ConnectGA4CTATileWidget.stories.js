/**
 * ConnectGA4CTATileWidget Component Stories.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import WithRegistrySetup from '../../../../../../tests/js/WithRegistrySetup';
import {
	provideModuleRegistrations,
	provideModules,
	provideSiteInfo,
	provideUserAuthentication,
	provideUserCapabilities,
} from '../../../../../../tests/js/utils';
import { withWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import ConnectGA4CTATileWidget from './ConnectGA4CTATileWidget';

const WidgetWithComponentProps = withWidgetComponentProps(
	'keyMetricsConnectAdSenseCTATile'
)( ConnectGA4CTATileWidget );

function Template() {
	return <WidgetWithComponentProps />;
}

export const Default = Template.bind( {} );
Default.storyName = 'ConnectGA4CTATileWidget';
Default.scenario = {
	label: 'KeyMetrics/ConnectGA4CTATileWidget',
	delay: 250,
};

export default {
	title: 'Key Metrics/ConnectGA4CTATileWidget',
	decorators: [
		( Story ) => {
			const setupRegistry = ( registry ) => {
				provideUserAuthentication( registry );
				provideUserCapabilities( registry );
				provideModules( registry, [
					{
						slug: 'analytics-4',
						active: false,
						connected: false,
					},
				] );
				provideModuleRegistrations( registry );
				provideSiteInfo( registry, {
					postTypes: [ { slug: 'product', label: 'Product' } ],
				} );
			};

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
