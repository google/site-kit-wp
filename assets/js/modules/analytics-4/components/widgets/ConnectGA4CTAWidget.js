/**
 * ConnectGA4CTA component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { useCallback, useEffect, useState } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { useSelect, useDispatch } from 'googlesitekit-data';
import { SpinnerButton } from 'googlesitekit-components';
import KeyMetricsCTAContent from '../../../../components/KeyMetrics/KeyMetricsCTAContent';
import KeyMetricsCTAFooter from '../../../../components/KeyMetrics/KeyMetricsCTAFooter';
import { CORE_MODULES } from '../../../../googlesitekit/modules/datastore/constants';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import { CORE_WIDGETS } from '../../../../googlesitekit/widgets/datastore/constants';
import { AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY } from '../../../../googlesitekit/widgets/default-areas';
import { MODULES_ANALYTICS_4 } from '../../datastore/constants';
import { CORE_LOCATION } from '../../../../googlesitekit/datastore/location/constants';
import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
import { KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY } from '../../constants';
import useActivateModuleCallback from '../../../../hooks/useActivateModuleCallback';
import useCompleteModuleActivationCallback from '../../../../hooks/useCompleteModuleActivationCallback';
import { useDebounce } from '../../../../hooks/useDebounce';

export default function ConnectGA4CTAWidget( { Widget, WidgetNull } ) {
	const ga4DependantKeyMetrics = useSelect( ( select ) => {
		const keyMetrics = select( CORE_USER ).getKeyMetrics();
		const widgets = select( CORE_WIDGETS ).getWidgets(
			AREA_MAIN_DASHBOARD_KEY_METRICS_PRIMARY
		);

		if ( ! keyMetrics || ! widgets ) {
			return [];
		}

		return widgets.filter(
			( { slug, modules } ) =>
				keyMetrics.includes( slug ) && modules.includes( 'analytics-4' )
		);
	} );
	const isAnalyticsActive = useSelect( ( select ) =>
		select( CORE_MODULES ).isModuleActive( 'analytics-4' )
	);
	const isNavigatingToReauthURL = useSelect( ( select ) => {
		const adminReauthURL =
			select( MODULES_ANALYTICS_4 ).getAdminReauthURL();

		if ( ! adminReauthURL ) {
			return false;
		}

		return select( CORE_LOCATION ).isNavigatingTo( adminReauthURL );
	} );
	const isActivatingAnalytics = useSelect( ( select ) =>
		select( CORE_MODULES ).isFetchingSetModuleActivation(
			'analytics-4',
			true
		)
	);
	const connectGA4URL = useSelect( ( select ) => {
		const settingsURL = select( CORE_SITE ).getAdminURL(
			'googlesitekit-settings'
		);

		return `${ settingsURL }#connected-services/analytics-4/edit`;
	} );
	const isNavigatingToGA4URL = useSelect( ( select ) =>
		select( CORE_LOCATION ).isNavigatingTo( connectGA4URL )
	);

	const { dismissItem } = useDispatch( CORE_USER );

	const activateAnalytics = useActivateModuleCallback( 'analytics-4' );
	const completeAnalyticsActivation =
		useCompleteModuleActivationCallback( 'analytics-4' );

	const handleCTAClick = useCallback( () => {
		if ( isAnalyticsActive ) {
			return completeAnalyticsActivation();
		}

		activateAnalytics();
	}, [ activateAnalytics, completeAnalyticsActivation, isAnalyticsActive ] );

	const [ inProgress, setInProgress ] = useState( false );

	/*
	 * Using debounce here because the spinner has to render across two separate calls.
	 * Rather than risk it flickering on and off in between the activation call completing and
	 * the navigate call starting, we will just set a debounce to keep the spinner for 3 seconds.
	 */
	const debouncedSetInProgress = useDebounce( setInProgress, 3000 );

	useEffect( () => {
		if (
			isActivatingAnalytics ||
			isNavigatingToReauthURL ||
			isNavigatingToGA4URL
		) {
			setInProgress( true );
		} else {
			debouncedSetInProgress( false );
		}
	}, [
		isActivatingAnalytics,
		isNavigatingToReauthURL,
		debouncedSetInProgress,
		isNavigatingToGA4URL,
	] );

	const isDismissed = useSelect( ( select ) =>
		select( CORE_USER ).isItemDismissed(
			KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY
		)
	);

	if ( isDismissed !== false || ga4DependantKeyMetrics.length < 4 ) {
		return <WidgetNull />;
	}

	return (
		<Widget
			noPadding
			Footer={ () => (
				<KeyMetricsCTAFooter
					onActionClick={ () =>
						dismissItem(
							KM_CONNECT_GA4_CTA_WIDGET_DISMISSED_ITEM_KEY
						)
					}
					showDismiss
				/>
			) }
		>
			<KeyMetricsCTAContent
				className="googlesitekit-km-connect-ga4-cta"
				title={ __( 'Analytics is disconnected', 'google-site-kit' ) }
				description={ __(
					'Metrics cannot be displayed without Analytics',
					'google-site-kit'
				) }
				actions={
					<SpinnerButton
						onClick={ handleCTAClick }
						isSaving={ inProgress }
						disabled={ inProgress }
					>
						{ __( 'Connect Analytics', 'google-site-kit' ) }
					</SpinnerButton>
				}
			/>
		</Widget>
	);
}
