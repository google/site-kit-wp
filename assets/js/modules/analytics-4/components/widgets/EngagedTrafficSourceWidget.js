/**
 * EngagedTrafficSourceWidget component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';
import { get } from 'lodash';

/**
 * WordPress dependencies
 */
import { __, sprintf } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { useSelect, useInViewSelect } from 'googlesitekit-data';
import {
	CORE_USER,
	KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
} from '../../../../googlesitekit/datastore/user/constants';
import {
	DATE_RANGE_OFFSET,
	MODULES_ANALYTICS_4,
} from '../../datastore/constants';
import { MetricTileText } from '../../../../components/KeyMetrics';
import { numFmt } from '../../../../util';
import ConnectGA4CTATileWidget from './ConnectGA4CTATileWidget';
import whenActive from '../../../../util/when-active';

function EngagedTrafficSourceWidget( props ) {
	const { Widget } = props;

	const dates = useSelect( ( select ) =>
		select( CORE_USER ).getDateRangeDates( {
			offsetDays: DATE_RANGE_OFFSET,
			compare: true,
		} )
	);

	const reportOptions = {
		...dates,
		dimensions: [ 'sessionDefaultChannelGroup' ],
		metrics: [ { name: 'engagedSessions' } ],
		orderBy: 'engagedSessions',
		limit: 1,
	};

	const report = useInViewSelect(
		( select ) => select( MODULES_ANALYTICS_4 ).getReport( reportOptions ),
		[ reportOptions ]
	);

	const error = useSelect( ( select ) =>
		select( MODULES_ANALYTICS_4 ).getErrorForSelector( 'getReport', [
			reportOptions,
		] )
	);

	const loading = useSelect(
		( select ) =>
			! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
				'getReport',
				[ reportOptions ]
			)
	);

	const { rows = [], totals = [] } = report || {};
	const makeFilter = ( dateRange, dimensionIndex ) => ( row ) =>
		get( row, `dimensionValues.${ dimensionIndex }.value` ) === dateRange;

	const topTrafficSource =
		rows.filter( makeFilter( 'date_range_0', 1 ) )[ 0 ]
			?.dimensionValues?.[ 0 ].value || '-';

	const currentEngagedSessions =
		parseFloat(
			rows.filter( makeFilter( 'date_range_0', 1 ) )[ 0 ]
				?.metricValues?.[ 0 ]?.value
		) || 0;
	const currentTotalEngagedSessions =
		parseFloat(
			totals.filter( makeFilter( 'date_range_0', 1 ) )[ 0 ]
				?.metricValues?.[ 0 ]?.value
		) || 0;
	const currentEngagedSessionsRate =
		currentEngagedSessions / currentTotalEngagedSessions || 0;

	const previousEngagedSessions =
		parseFloat(
			rows.filter( makeFilter( 'date_range_1', 1 ) )[ 0 ]
				?.metricValues?.[ 0 ]?.value
		) || 0;
	const previousTotalEngagedSessions =
		parseFloat(
			totals.filter( makeFilter( 'date_range_1', 1 ) )[ 0 ]
				?.metricValues?.[ 0 ]?.value
		) || 0;
	const previousEngagedSessionsRate =
		previousEngagedSessions / previousTotalEngagedSessions || 0;

	const format = {
		style: 'percent',
		signDisplay: 'never',
		maximumFractionDigits: 1,
	};

	return (
		<MetricTileText
			Widget={ Widget }
			widgetSlug={ KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE }
			metricValue={ topTrafficSource }
			metricValueFormat={ format }
			subText={ sprintf(
				/* translators: 1. Percentage of total engaged sessions. 2: Total number of engaged sessions. */
				__( '%1$s of %2$s engaged sessions', 'google-site-kit' ),
				numFmt( currentEngagedSessionsRate, format ),
				numFmt( currentTotalEngagedSessions, { style: 'decimal' } )
			) }
			previousValue={ previousEngagedSessionsRate }
			currentValue={ currentEngagedSessionsRate }
			loading={ loading }
			error={ error }
			moduleSlug="analytics-4"
		/>
	);
}

EngagedTrafficSourceWidget.propTypes = {
	Widget: PropTypes.elementType.isRequired,
};

export default whenActive( {
	moduleName: 'analytics-4',
	FallbackComponent: ConnectGA4CTATileWidget,
} )( EngagedTrafficSourceWidget );
