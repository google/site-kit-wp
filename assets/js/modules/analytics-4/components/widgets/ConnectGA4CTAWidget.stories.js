/**
 * ConnectGA4CTAWidget Component Stories.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
	KM_ANALYTICS_RETURNING_VISITORS,
	KM_ANALYTICS_NEW_VISITORS,
	KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
} from '../../../../googlesitekit/datastore/user/constants';
import { provideKeyMetrics } from '../../../../../../tests/js/utils';
import { provideKeyMetricsWidgetRegistrations } from '../../../../components/KeyMetrics/test-utils';
import { withWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import WithRegistrySetup from '../../../../../../tests/js/WithRegistrySetup';
import ConnectGA4CTAWidget from './ConnectGA4CTAWidget';

const WidgetWithComponentProps = withWidgetComponentProps(
	'keyMetricsConnectGA4CTA'
)( ConnectGA4CTAWidget );

function Template() {
	return (
		<div className="googlesitekit-widget-area--mainDashboardKeyMetricsPrimary">
			<div className="googlesitekit-widget--keyMetricsConnectGA4All">
				<WidgetWithComponentProps />
			</div>
		</div>
	);
}

export const Default = Template.bind( {} );
Default.storyName = 'ConnectGA4CTAWidget';
Default.scenario = {
	label: 'KeyMetrics/ConnectGA4CTAWidget',
};

export default {
	title: 'Key Metrics/ConnectGA4CTAWidget',
	decorators: [
		( Story ) => {
			const setupRegistry = ( registry ) => {
				const keyMetricWidgets = [
					KM_ANALYTICS_RETURNING_VISITORS,
					KM_ANALYTICS_NEW_VISITORS,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
					KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
				];

				provideKeyMetrics( registry, {
					widgetSlugs: keyMetricWidgets,
				} );

				provideKeyMetricsWidgetRegistrations(
					registry,
					keyMetricWidgets.reduce(
						( acc, widget ) => ( {
							...acc,
							[ widget ]: { modules: [ 'analytics-4' ] },
						} ),
						{}
					)
				);
			};

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
