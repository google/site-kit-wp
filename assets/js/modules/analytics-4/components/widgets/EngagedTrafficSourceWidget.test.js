/**
 * EngagedTrafficSourceWidget component tests.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { render } from '../../../../../../tests/js/test-utils';
import {
	createTestRegistry,
	freezeFetch,
	provideKeyMetrics,
	provideModules,
} from '../../../../../../tests/js/utils';
import { provideAnalytics4MockReport } from '../../utils/data-mock';
import { getWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
import { withConnected } from '../../../../googlesitekit/modules/datastore/__fixtures__';
import {
	CORE_USER,
	KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
} from '../../../../googlesitekit/datastore/user/constants';
import { DATE_RANGE_OFFSET } from '../../datastore/constants';
import { ERROR_REASON_INSUFFICIENT_PERMISSIONS } from '../../../../util/errors';
import EngagedTrafficSourceWidget from './EngagedTrafficSourceWidget';

describe( 'EngagedTrafficSourceWidget', () => {
	let registry;
	const { Widget, WidgetNull } = getWidgetComponentProps(
		KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE
	);
	const reportEndpoint = new RegExp(
		'^/google-site-kit/v1/modules/analytics-4/data/report'
	);

	beforeEach( () => {
		registry = createTestRegistry();
		registry.dispatch( CORE_USER ).setReferenceDate( '2020-09-08' );
		provideKeyMetrics( registry );
		provideModules( registry, withConnected( 'analytics-4' ) );
	} );

	it( 'should render correctly with the expected metrics', async () => {
		provideAnalytics4MockReport( registry, {
			...registry.select( CORE_USER ).getDateRangeDates( {
				offsetDays: DATE_RANGE_OFFSET,
				compare: true,
			} ),
			dimensions: [ 'sessionDefaultChannelGroup' ],
			metrics: [ { name: 'engagedSessions' } ],
			orderBy: 'engagedSessions',
			limit: 1,
		} );

		const { container, waitForRegistry } = render(
			<EngagedTrafficSourceWidget
				Widget={ Widget }
				WidgetNull={ WidgetNull }
			/>,
			{ registry }
		);
		await waitForRegistry();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render the loading state while resolving the report', async () => {
		// Freeze the report fetch to keep the widget in loading state.
		freezeFetch( reportEndpoint );

		const { container, waitForRegistry } = render(
			<EngagedTrafficSourceWidget
				Widget={ Widget }
				WidgetNull={ WidgetNull }
			/>,
			{ registry }
		);
		await waitForRegistry();

		[
			'.googlesitekit-km-widget-tile__loading',
			'.googlesitekit-km-widget-tile__loading-header',
			'.googlesitekit-km-widget-tile__loading-body',
		].forEach( ( selector ) => {
			expect( container.querySelector( selector ) ).toBeInTheDocument();
		} );

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render the insufficient permissions error variant when the report fetch fails', async () => {
		const errorResponse = {
			code: 'test_error',
			message: 'Error message.',
			data: { reason: ERROR_REASON_INSUFFICIENT_PERMISSIONS },
		};

		fetchMock.get( reportEndpoint, {
			body: errorResponse,
			status: 500,
		} );

		const { container, getByText, waitForRegistry } = render(
			<EngagedTrafficSourceWidget
				Widget={ Widget }
				WidgetNull={ WidgetNull }
			/>,
			{ registry }
		);

		await waitForRegistry();

		expect( console ).toHaveErrored();

		expect(
			container.querySelector( '.googlesitekit-km-widget-tile--error' )
		).toBeInTheDocument();

		expect( getByText( /Insufficient permissions/i ) ).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );
} );
