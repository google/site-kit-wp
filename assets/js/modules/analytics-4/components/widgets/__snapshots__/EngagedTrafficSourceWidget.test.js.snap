// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EngagedTrafficSourceWidget should render correctly with the expected metrics 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--kmAnalyticsEngagedTrafficSource googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-km-widget-tile googlesitekit-km-widget-tile--text"
      >
        <div
          class="googlesitekit-km-widget-tile__title-container"
        >
          <h3
            class="googlesitekit-km-widget-tile__title"
          >
            Most engaged traffic source
          </h3>
          <span
            class="googlesitekit-info-tooltip"
            title="Channel (e.g. social, paid, search) that brought in the most visitors who had a meaningful engagement with your site"
          >
            <svg />
          </span>
        </div>
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-widget-tile__metric-container"
          >
            <div
              class="googlesitekit-km-widget-tile__metric"
            >
              Organic Search
            </div>
            <p
              class="googlesitekit-km-widget-tile__subtext"
            >
              100% of 55 engaged sessions
            </p>
          </div>
          <div
            class="googlesitekit-km-widget-tile__metric-change-container"
          >
            <div
              class="googlesitekit-change-badge googlesitekit-change-badge--zero"
            >
              0%
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EngagedTrafficSourceWidget should render the generic error variant when the report fetch fails 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-cta__header_text"
        >
          Most engaged traffic source
        </h2>
        <span
          class="googlesitekit-info-tooltip"
          title="Channel (e.g. social, paid, search) that brought in the most visitors who had a meaningful engagement with your site"
        >
          <svg />
        </span>
      </div>
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-cta__title"
        >
          Data loading failed
        </h3>
        
        
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <div
            class="googlesitekit-error-retry-text"
          >
            <a
              aria-label="Get help (opens in a new tab)"
              class="googlesitekit-cta-link"
              href="undefined?error_id=internal_server_error"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Get help
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EngagedTrafficSourceWidget should render the insufficient permissions error variant when the report fetch fails 1`] = `
<div>
  <div
    class="googlesitekit-km-widget-tile--error"
  >
    <div
      class="googlesitekit-cta googlesitekit-cta--error"
    >
      <div
        class="googlesitekit-cta__header"
      >
        <h2
          class="googlesitekit-cta__header_text"
        >
          Most engaged traffic source
        </h2>
        <span
          class="googlesitekit-info-tooltip"
          title="Channel (e.g. social, paid, search) that brought in the most visitors who had a meaningful engagement with your site"
        >
          <svg />
        </span>
      </div>
      <div
        class="googlesitekit-cta__body"
      >
        <h3
          class="googlesitekit-cta__title"
        >
          Insufficient permissions
        </h3>
        
        
        
        
        <div
          class="googlesitekit-report-error-actions"
        >
          <div
            class="googlesitekit-error-retry-text"
          >
            Trouble getting access? 
            <a
              aria-label="Get help (opens in a new tab)"
              class="googlesitekit-cta-link"
              href="undefined?error_id=analytics-4_insufficient_permissions"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="googlesitekit-cta-link__contents"
              >
                Get help
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EngagedTrafficSourceWidget should render the loading state while resolving the report 1`] = `
<div>
  <div
    class="googlesitekit-widget googlesitekit-widget--kmAnalyticsEngagedTrafficSource googlesitekit-widget--no-padding"
  >
    <div
      class="googlesitekit-widget__body"
    >
      <div
        class="googlesitekit-km-widget-tile googlesitekit-km-widget-tile--text"
      >
        <div
          class="googlesitekit-km-widget-tile__title-container"
        >
          <h3
            class="googlesitekit-km-widget-tile__title"
          >
            Most engaged traffic source
          </h3>
          <span
            class="screen-reader-text"
          >
            <span
              class="googlesitekit-info-tooltip"
              title="Channel (e.g. social, paid, search) that brought in the most visitors who had a meaningful engagement with your site"
            >
              <svg />
            </span>
          </span>
        </div>
        <div
          class="googlesitekit-km-widget-tile__body"
        >
          <div
            class="googlesitekit-km-widget-tile__loading"
          >
            <div
              class="googlesitekit-preview-block googlesitekit-km-widget-tile__loading-header"
              style="width: 100%; height: 14px;"
            >
              <div
                class="googlesitekit-preview-block__wrapper"
              />
            </div>
            <div
              class="googlesitekit-preview-block googlesitekit-km-widget-tile__loading-body"
              style="width: 100%; height: 53px;"
            >
              <div
                class="googlesitekit-preview-block__wrapper"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
